### ISSUE TITLE
About Training Iteration Speed

### ISSUE THREAD

**Issue by steinate** [NIRVANALAN/STream3R#18]
My training setup is 8 GPUs, same with setting reported in the paper, and I have not modified any configuration files of the project. However, the current iteration speed on my side is only 0.23 it/s. I am wondering if this speed is normal?

**Comment by Luo-Yihang** [comment]
Hi, the training speed looks normal. I remember it being around 0.25 - 0.35 it/s on A100/H200.

Regarding the requirements.txt, could you please let me know if any dependencies are missing? I’ve also provided requirements file with the exact versions I used below. Note that the environment listed below may include some unnecessary dependencies.


```
absl-py==1.4.0
accelerate==1.11.0
addict==2.4.0
aiofiles==24.1.0
aiohappyeyeballs==2.6.1
aiohttp==3.12.15
aiohttp-cors==0.8.1
aiosignal==1.4.0
albucore==0.0.24
albumentations==2.0.8
alembic==1.16.5
annotated-types==0.7.0
antlr4-python3-runtime==4.9.3
anyio==4.10.0
appdirs==1.4.4
argcomplete==3.6.3
argon2-cffi==25.1.0
argon2-cffi-bindings==25.1.0
array_record==0.8.1
arrow==1.3.0
asttokens==3.0.0
astunparse==1.6.3
async-lru==2.0.5
attrs==25.3.0
autopage==0.5.2
babel==2.17.0
beautifulsoup4==4.13.5
bleach==6.2.0
blinker==1.9.0
boto3==1.40.50
botocore==1.40.50
Brotli==1.1.0
cachetools==5.5.2
certifi==2025.8.3
cffi==2.0.0
cfgv==3.4.0
charset-normalizer==3.4.3
chex==0.1.91
click==8.3.0
cliff==4.11.0
cloudpickle==3.1.2
cmaes==0.12.0
cmd2==2.7.0
colorama==0.4.6
coloredlogs==15.0.1
colorful==0.5.7
colorlog==6.9.0
colorspacious==1.1.2
comm==0.2.3
ConfigArgParse==1.7.1
contourpy==1.3.3
cycler==0.12.1
dacite==1.8.1
dash==3.2.0
dask==2023.3.1
debugpy==1.8.17
decorator==4.4.2
decord==0.6.0
deepspeed==0.16.7
defusedxml==0.7.1
distlib==0.4.0
dm-haiku==0.0.15
dm-tree==0.1.9
docstring_parser==0.17.0
easydict==1.13
einops==0.8.1
einshape==1.0
einsum==0.3.0
einx==0.3.0
embreex==2.17.7.post6
environs==14.3.0
etils==1.13.0
evo==1.33.0
executing==2.2.1
fastapi==0.116.2
fastjsonschema==2.21.2
ffmpy==0.6.1
filelock==3.13.1
Flask==3.1.2
flatbuffers==25.9.23
flax==0.11.2
flow-vis==0.1
fonttools==4.60.0
fqdn==1.5.1
frozendict==2.4.6
frozenlist==1.7.0
fsspec==2024.6.1
gast==0.4.0
gdown==5.2.0
gitdb==4.0.12
GitPython==3.1.45
glcontext==3.0.0
google-api-core==2.25.1
google-auth==2.41.0
google-pasta==0.2.0
googleapis-common-protos==1.70.0
gradio==5.31.0
gradio_client==1.10.1
greenlet==3.2.4
groovy==0.1.2
grpcio==1.74.0
h11==0.16.0
h5py==3.14.0
hf-xet==1.2.0
hjson==3.1.0
httpcore==1.0.9
httpx==0.28.1
huggingface_hub==1.3.2
humanfriendly==10.0
humanize==4.13.0
hydra-colorlog==1.2.0
hydra-core==1.3.2
hydra-optuna-sweeper==1.2.0
identify==2.6.14
idna==3.10
imageio==2.37.0
imageio-ffmpeg==0.6.0
immutabledict==2.2.0
importlib_metadata==8.7.0
importlib_resources==6.5.2
iniconfig==2.1.0
iopath==0.1.10
ipdb==0.13.13
ipykernel==6.30.1
ipympl==0.9.7
ipython==9.5.0
ipython_pygments_lexers==1.1.1
ipywidgets==8.1.7
isoduration==20.11.0
itsdangerous==2.2.0
jax==0.7.2
jaxlib==0.7.2
jaxline==0.0.8
jaxtyping==0.3.2
jedi==0.19.2
Jinja2==3.1.4
jmespath==1.0.1
jmp==0.0.4
joblib==1.5.2
json5==0.12.1
jsonpointer==3.0.0
jsonschema==4.25.1
jsonschema-specifications==2025.9.1
jupyter-events==0.12.0
jupyter-lsp==2.3.0
jupyter_client==8.6.3
jupyter_core==5.8.1
jupyter_server==2.17.0
jupyter_server_terminals==0.5.3
jupyterlab==4.4.7
jupyterlab_pygments==0.3.0
jupyterlab_server==2.27.3
jupyterlab_widgets==3.0.15
keras==2.13.1
kiwisolver==1.4.9
kornia==0.8.1
kornia_rs==0.1.9
lark==1.2.2
lazy_loader==0.4
libclang==18.1.1
lightning==2.5.5
lightning-bolts==0.7.0
lightning-utilities==0.15.2
locket==1.0.0
lxml==6.0.1
lz4==4.4.4
Mako==1.3.10
manifold3d==3.2.1
mapbox_earcut==1.0.3
Markdown==3.9
markdown-it-py==4.0.0
MarkupSafe==2.1.5
marshmallow==4.0.1
matplotlib==3.10.6
matplotlib-inline==0.1.7
mdurl==0.1.2
mediapy==1.2.4
mistune==3.1.4
ml_collections==1.1.0
ml_dtypes==0.5.3
moderngl==5.12.0
moviepy==1.0.0
mpmath==1.3.0
msgpack==1.1.1
msgspec==0.19.0
multidict==6.6.4
multiprocessing-logging==0.3.4
namex==0.1.0
narwhals==2.5.0
natsort==8.4.0
nbclient==0.10.2
nbconvert==7.16.6
nbformat==5.10.4
nest-asyncio==1.6.0
networkx==3.3
ninja==1.13.0
nodeenv==1.9.1
notebook_shim==0.2.4
numexpr==2.14.1
numpy==2.4.0
nvidia-cublas-cu12==12.4.5.8
nvidia-cuda-cupti-cu12==12.4.127
nvidia-cuda-nvrtc-cu12==12.4.127
nvidia-cuda-runtime-cu12==12.4.127
nvidia-cudnn-cu12==9.1.0.70
nvidia-cufft-cu12==11.2.1.3
nvidia-curand-cu12==10.3.5.147
nvidia-cusolver-cu12==11.6.1.9
nvidia-cusparse-cu12==12.3.1.170
nvidia-cusparselt-cu12==0.6.2
nvidia-ml-py==13.580.82
nvidia-nccl-cu12==2.21.5
nvidia-nvjitlink-cu12==12.4.127
nvidia-nvtx-cu12==12.4.127
nvitop==1.5.3
oauthlib==3.3.1
omegaconf==2.3.0
open3d==0.19.0
opencensus==0.11.4
opencensus-context==0.1.3
opencv-python==4.11.0.86
opencv-python-headless==4.12.0.88
opentelemetry-api==1.37.0
opentelemetry-exporter-prometheus==0.58b0
opentelemetry-proto==1.37.0
opentelemetry-sdk==1.37.0
opentelemetry-semantic-conventions==0.58b0
opt_einsum==3.4.0
optax==0.2.6
optree==0.17.0
optuna==2.10.1
orbax-checkpoint==0.11.25
orjson==3.11.3
overrides==7.7.0
packaging==25.0
pako==0.3.1
pandas==2.3.2
pandocfilters==1.5.1
parso==0.8.5
partd==1.4.2
pause==0.3
petrel-oss-sdk==2.2.2.post3
pexpect==4.9.0
Pillow==9.2.0
pillow_heif==1.1.0
platformdirs==4.4.0
plotly==5.13.1
pluggy==1.6.0
plyfile==1.1.2
portalocker==3.2.0
pre_commit==4.3.0
prettytable==3.16.0
proglog==0.1.12
prometheus_client==0.23.1
promise==2.3
prompt_toolkit==3.0.52
propcache==0.3.2
proto-plus==1.26.1
protobuf==4.25.8
psutil==5.9.8
ptyprocess==0.7.0
pure_eval==0.2.3
py-cpuinfo==9.0.0
py-spy==0.4.1
pyarrow==16.0.0
pyasn1==0.6.1
pyasn1_modules==0.4.2
pyceres==2.4
pycollada==0.9.2
pycolmap==3.11.1
pycparser==2.23
pydantic==2.11.9
pydantic_core==2.33.2
pydub==0.25.1
pyglet==1.5.31
Pygments==2.19.2
PyJWT==2.10.1
pyparsing==3.2.4
pyperclip==1.10.0
pyquaternion==0.9.9
pyro-api==0.1.2
pyro-ppl==1.9.1
PySocks==1.7.1
pytest==8.4.2
python-dateutil==2.9.0.post0
python-dotenv==1.1.1
python-json-logger==3.3.0
python-multipart==0.0.20
pytorch-lightning==1.9.5
pytorch3d @ git+https://github.com/facebookresearch/pytorch3d.git@75ebeeaea0908c5527e7b1e305fbc7681382db47
pytz==2025.2
PyWavelets==1.8.0
PyYAML==6.0.2
pyzmq==27.1.0
ray==2.49.2
referencing==0.36.2
regex==2025.9.18
requests==2.32.5
requests-oauthlib==2.0.0
retrying==1.4.2
rfc3339-validator==0.1.4
rfc3986-validator==0.1.1
rfc3987-syntax==1.1.0
rich==14.1.0
rich-argparse==1.7.1
roma==1.5.4
rootutils==1.0.7
rosbags==0.11.0
rpds-py==0.27.1
rsa==4.9.1
rtree==1.4.1
ruamel.yaml==0.18.16
ruamel.yaml.clib==0.2.14
ruff==0.13.1
s3transfer==0.14.0
safehttpx==0.1.6
safetensors==0.6.2
scenepic==1.1.1
scikit-image==0.25.2
scikit-learn==1.7.2
scipy==1.15.3
seaborn==0.13.2
segment_anything @ git+https://github.com/facebookresearch/segment-anything.git@dca509fe793f601edb92606367a655c15ac00fdf
semantic-version==2.10.0
Send2Trash==1.8.3
sentry-sdk==2.38.0
shapely==2.1.1
shellingham==1.5.4
simple-parsing==0.1.7
simplejson==3.20.1
simsimd==6.5.3
six==1.17.0
smart_open==7.3.1
smmap==5.0.2
sniffio==1.3.1
soupsieve==2.8
spaces==0.42.1
SQLAlchemy==2.0.43
stack-data==0.6.3
starlette==0.48.0
stevedore==5.5.0
stringzilla==4.0.14
svg.path==7.0
sympy==1.13.1
tabulate==0.9.0
tapnet @ file:///root/Projects/tapnet
tenacity==9.1.2
tensorboard==2.20.0
tensorboard-data-server==0.7.2
tensorflow==2.20.0
tensorflow-datasets==4.9.9
tensorflow-estimator==2.13.0
tensorflow-io-gcs-filesystem==0.37.1
tensorflow-metadata==1.17.2
tensorstore==0.1.77
termcolor==3.1.0
terminado==0.18.1
threadpoolctl==3.6.0
tifffile==2025.9.9
timm==1.0.20
tinycss2==1.4.0
tokenizers==0.22.1
toml==0.10.2
tomlkit==0.13.3
toolz==1.0.0
torch==2.6.0+cu124
torchinfo==1.8.0
torchmetrics==1.8.2
torchvision==0.21.0+cu124
tornado==6.5.2
tqdm==4.67.1
traitlets==5.14.3
transformers==5.0.0rc1
treescope==0.1.10
trimesh==4.8.2
triton==3.2.0
typer==0.17.4
typer-slim==0.21.1
types-python-dateutil==2.9.0.20250822
typing-inspection==0.4.1
typing_extensions==4.15.0
tzdata==2025.2
uri-template==1.3.0
urllib3==2.5.0
utils3d @ git+https://github.com/EasternJournalist/utils3d.git@d3a577acf0a9ad7e513a1416449a07b6f47d967f
uvicorn==0.35.0
vhacdx==0.0.8.post2
virtualenv==20.34.0
viser==1.0.10
wadler_lindig==0.1.7
wandb==0.22.0
wcwidth==0.2.13
webcolors==24.11.1
webencodings==0.5.1
websocket-client==1.8.0
websockets==15.0.1
Werkzeug==3.1.3
widgetsnbextension==4.0.14
wrapt==1.17.3
xformers==0.0.29.post3
xxhash==3.5.0
yarl==1.20.1
yourdfpy==0.0.58
zipp==3.23.0
zstandard==0.25.0
```

**Comment by steinate** [comment]
Think you! `botocore` maybe missing in the requirements.txt.

**Comment by junzastar** [comment]
Hi, I try to launch the training process on two 4090 GPUs. But the process hangs at the beginning or after several iterations (nccl times out). 
Do you have any advice? Thank you!

**Comment by Luo-Yihang** [comment]
Hi @junzastar, firstly, you can check whether all the data are sensible, for example, whether the there is GT supervision corresponds to all outputs for all data?

If everything looks good, one way to bypass this issue is to add:

`+(pred_pts3d_global * 0).mean() + (pred_depth * 0).mean() + (pred_pose_enc_list[-1] * 0).mean()`

after [this line](https://github.com/NIRVANALAN/STream3R/blob/9d6fcf2472aa21752ffffb27206be1dc3fd4142f/stream3r/loss/losses.py#L252).
